### PR TITLE
Added Bernoulli Naive Bayes

### PR DESCRIPTION
**Pull request by Sar-thak-3**

### CODE REVIEW

File: src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                    numerator++;
+                }
+            }
+        }
+        return numerator/double(denominator);
+    }
+
+template<class T>
+    int predict(std::vector<std::vector<T>> x_train,std::vector<T> y_train,std::vector<T> x_test){
+        int n_features = x_train[0].size();
```
Comment (ken1000minus7) [inline src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.cpp:36]: According to our conventions, variable and function names should be in camel case so make these changes wherever required

File: src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.hpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+template<class T>
+class BernoulliNB{  
+private:
+    double prior_prob(std::vector<T> y_train,int label);
+    double conditional_prob(std::vector<std::vector<T>> x_train,std::vector<T> y_train,int feature_col,int feature_val,int label);
+
+public:
+    int predict(std::vector<std::vector<T>> x_train,std::vector<T> y_train,std::vector<T> x_test);
+};
+
```
Comment (ken1000minus7) [inline src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.hpp:21]: Add doc comments above each explaining the parameters and what it returns, refer other files for the exact format

**Review by ken1000minus7 — CHANGES_REQUESTED** [review]
Small changes needed to meet the conventions

File: src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        int denominator=0,numerator=0;
+        for(int i=0;i<y_train.size();i++){
+            if(y_train[i]==label){
+                denominator++;
+                if(x_train[i][feature_col]==feature_val){
+                    numerator++;
+                }
+            }
+        }
+        return numerator/double(denominator);
```
Comment (uttammittal02) [inline src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.cpp:31]: Add an check for denominator != 0

**Review by uttammittal02 — CHANGES_REQUESTED** [review]
Add a fit function to train the model

**Comment by ken1000minus7** [comment]
Change `predict_prob` to camel case as well

File: src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.cpp [hunk 4]
```diff
@@ -6,57 +6,61 @@
...
+            return numerator/double(denominator);
+        }
+        return numerator;
+        
     }
 
 template<class T>
-    int predict(std::vector<std::vector<T>> x_train,std::vector<T> y_train,std::vector<T> x_test){
-        int n_features = x_train[0].size();
+    int fit(std::vector<std::vector<T>> xTrain,std::vector<T> yTrain,std::vector<T> xTest){
```
Comment (uttammittal02) [inline src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.cpp:39]: You need to make 2 different functions for fit and predict. Also make a function fit_predict...basically the current fit function

**Review by uttammittal02 — CHANGES_REQUESTED** [review]

File: src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
+/**
+ * @file methods/neighbors/bernoulli_nb/nernoulli_nb.cpp
+ *
+ * Implementation of the Bernoulli Naive Bayes main program
+ */
+#include "bernoulli_nb.hpp"
+
+template<class T>
+    double prior_prob(std::vector<T> yTrain,int label){   // Prior-Probability P(y)
```
Comment (uttammittal02) [inline src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.cpp:9]: Change to camel case

File: src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.cpp [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        for(int i=0;i<yTrain.size();i++){
+            if(yTrain[i]==label){
+                sum += 1;
+            }
+        }
+        return sum/double(yTrain.size());
+    }
+
+template<class T>
+    double conditional_prob(std::vector<std::vector<T>> xTrain,std::vector<T> yTrain,int featureCol,int featureVal,int label){ // Conditional Probability
```
Comment (uttammittal02) [inline src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.cpp:20]: Change to camel case

**Review by uttammittal02 — CHANGES_REQUESTED** [review]

File: src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.hpp (hunk 2 above)
Comment (uttammittal02) [reply src/slowmokit/methods/neighbors/bernoulli_nb/bernoulli_nb.hpp]: > Add doc comments above each function explaining the parameters and what it returns, refer other files for the exact format

@Sar-thak-3 this also needs to be done properly. See [this file](https://github.com/PEC-CSS/slowmokit/blob/main/src/slowmokit/methods/metrics/accuracy.hpp) for reference

**Comment by ken1000minus7** [comment]
Hey @Sar-thak-3 , your PR is almost ready to be merged, the need to add above the functions should be doc comments
https://github.com/PEC-CSS/slowmokit/blob/08e2d497cbf1dcf8516570cbead34c5cd2002e3a/src/slowmokit/ducks/io/io.hpp#L15-L20
Here is an example of how the comment should look like

**Comment by ken1000minus7** [comment]
@Sar-thak-3 Add two stars in the beginning
With a single star it is a simple multiline comment
```cpp
/* 

 */
```
With two stars it is marked as a doc comment
```cpp
/** 

*/
```
You need to write params by first writing `@param` then name of variable then it's description, don't write write colons after`@param`
example
```cpp
/** 
*  @param x it is a variable
*/
```
Also separate params should be written in separate lines, like this
```cpp
/** 
*  @param x it is a variable
*  @param y it is another variable
*/
```
Rest all is fine, just make these changes

**Review by ken1000minus7 — APPROVED** [review]
Perfect, great work

**Review by uttammittal02 — APPROVED** [review]
Good job

**Pull request merged**